### PR TITLE
feat(helm chart): network policies are provided for all services

### PR DESCRIPTION
**Pull request by iainsproat**
## Description & motivation

[Network policies](https://kubernetes.io/docs/concepts/services-networking/network-policies/) are used to deny arbitrary network egress and ingress to a pod.  They effectively act as a simple 'firewall' around each service.   This provides some additional security at the kubernetes network-mesh level, for example preventing an attacker connecting to other pods within the cluster or exfiltrating data to arbitrary locations.

Fixes https://github.com/specklesystems/speckle-server/issues/860

## Changes:

Frontend:
- only allow ingress to port 80 from within the cluster (which should be the ingress loadbalancer)
- block all egress

Backend:
- only allow ingress to port 3000 from within the cluster (which should be the ingress loadbalancer)
- explicitly allow ingress from test and file import service.
- allow egress to redis, postgres, & blob storage

Preview service
- only allow ingress from the monitoring stack to the monitoring port
- allow egress to postgres

Monitoring
- only allow ingress from the monitoring stack to the monitoring port
- allow egress to postgres

File Import Service
- only allow ingress from the monitoring stack to the monitoring port
- allow egress to postgres
- allow egress to the speckle-server (backend) port 3000

Test
- block all ingress
- only allow egress to the speckle-server (backend) port 3000

Webhook service
- only allow ingress from the monitoring stack to the monitoring port
- only allow egress to outside of the cluster.  This assumes that applications called by the webhook service are not within the cluster (or are accessible via an ingress from the internet).  This prevents the webhook service being abused to call other kubernetes services within the cluster.

Screenshot 2022-08-15 at 10 47 54

_diagram of the configured ingress (red/purple) and egress (greens), and both ingress/egress (black)_

## To-do before merge

No special requirements

## Screenshots:

See comment below.

## Validation of changes:

See comment below.

## Things that aren't perfect

- Kubernetes Network Policies don't support domain names, only IP addresses.  Hence if an endpoint (e.g. redis or postgres) is a domain name, and we don't know the IP address will be static (very likely), we have to allow egress to everywhere 0.0.0.0/0.  Separate issue to address this for Cilium network meshes at some point in the future: https://github.com/specklesystems/speckle-server/issues/913
- Kubernetes Network Policies only support IP v4 CIDRs.
- I don't think it's possible to explicitly select on the nginx ingress, so ingress to the frontend and server is allowed from anywhere in the cluster.

## Checklist:

- [x] My pull request follows the guidelines in the [Contributing guide](https://github.com/specklesystems/speckle-server/blob/main/CONTRIBUTING.md)?
- [x] My pull request does not duplicate any other open [Pull Requests](../../pulls) for the same update/change?
- [x] My commits are related to the pull request and do not amend unrelated code or documentation.
- [x] My code follows a similar style to existing code.
- [x] I have added appropriate tests. - was deployed to a cluster, see screenshots below.
- [x] I have updated or added relevant documentation.

### CODE REVIEW

**Comment by iainsproat** [comment]
## Proof of work

- No policies are deployed when `.values.*.networkPolicy.enabled` is `false`:
Screenshot 2022-08-10 at 15 56 47

- Policies are deployed when `.values.*.networkPolicy.enabled` is `true` (default):
Screenshot 2022-08-10 at 16 09 38

File: utils/helm/speckle-server/templates/fileimport_service/networkpolicy.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      ports:
+        - port: 53
+          protocol: UDP
+    # allow egress to speckle-server
+    - to:
+        - podSelector:
+            matchLabels:
+{{ include "server.selectorLabels" $ | indent 16 }}
+      ports:
+        - port: 3000
```
Comment (iainsproat) [inline utils/helm/speckle-server/templates/fileimport_service/networkpolicy.yml:40]: Tried using named port `http`, but this wasn't working 🤔

File: utils/helm/speckle-server/templates/_helpers.tpl [hunk 2]
```diff
@@ -93,3 +93,137 @@ Part-of label
...
+Creates a network policy egress definition for connecting to Postgres
+*/}}
+{{- define "speckle.networkpolicy.egress.blob_storage" -}}
+{{- $port := (default "443" .Values.s3.networkPolicy.port ) -}}
+{{- if .Values.s3.networkPolicy.inCluster.enabled -}}
+{{ include "speckle.networkpolicy.egress.internal" (dict "podSelector" .Values.s3.networkPolicy.inCluster.podSelector "namespaceSelector" .Values.s3.networkPolicy.inCluster.namespaceSelector "port" $port) }}
+{{- else if .Values.s3.networkPolicy.externalToCluster.enabled -}}
+  {{- $host := (urlParse .Values.s3.endpoint).host -}}
+  {{- $ip := "" -}}
+  {{- if eq (include "speckle.isIPv4" $host) "true" -}}
```
Comment (iainsproat) [inline utils/helm/speckle-server/templates/_helpers.tpl:141]: Struggled to have the include return a boolean (i.e. not a string) and avoid using `eq` against a string value of "true".  Kept failing for some reason I was unable to debug.

**Comment by gjedlicska** [comment]
2 things are missing I think:
- [ ] allow egress to sentry from those services which report stuff (just the server afaik)
- [ ] allow egress to mixpanel for metrics

**Comment by gjedlicska** [comment]
what's the difference between the colors of egress lines?

**Comment by iainsproat** [comment]
@gjedlicska - the shades of green are just to make it easier to trace the line.
greens are egress only. purples are ingress only, and black is both ingress/egress on the respective service (e.g. egress for fileimport, ingress for server).

**Comment by iainsproat** [comment]
@gjedlicska - sentry_dns value was only ingested by the server, so I assume that is the only service that was using it.

Mixpanel is only referenced in the frontend code, and my understanding is that mixpanel runs in the user's browser.  i.e. there's no need for egress from frontend.

Fixed the services with unnecessary access to s3 and redis.

Diagram is updated.

**Review by gjedlicska — APPROVED** [review]

**Pull request merged**